Week 05 controlled components forms

savir-nosaj opened the pull request:
### Changes =>

**App.jsx**
* correcting prop syntax in handleAddTodo function
* create new function that will set isCompleted prop to true on form click from child component
* pass new function as prop to ToDoList

**index.css**
* change to Dark Mode because of eye strain

**ToDoForm.jsx**
* create new string state (will contain input inserted to input)
* made edits to handleAddTodoTitle:
1. adding additional attributes to input (value: set to workingTodoTitle state, onChange: append input value to workingTodoTitle state)
2. adding disabled button functionally using logic

**ToDoList.jsx**
* destructure onCompletedTodo prop
* create new variable (filteredTodoList containing a filter, of all items with a isCompleted prop set to true
* display data depending on filter
* figured that passing key prop to ToDoListItem child is unnecessary since it’s already stored in todoList prop.
* similar with todo; there’s no need to pass specifically title. pass the entire todoList prop to child

**ToDoListItem.jsx**
* destructure onCompletedTodo prop
* li now holds a form, with checkbox, with onChange event-listener that calls onCompletedTodo, passing item key id

savir-nosaj committed:
Update App.jsx

changes =>

- correcting prop syntax in handleAddTodo function
- create new function that will set isCompleted prop to true on form click from child component
- pass new function as prop to ToDoList

savir-nosaj committed:
Update index.css

eye strain (asthenopia)

savir-nosaj committed:
Update ToDoForm.jsx

* create new string state (will contain input inserted to input)
* made edits to handleAddTodoTitle
1. adding additional attributes to input (value: set to workingTodoTitle state, onChange: append input value to workingTodoTitle state)
2. 
* adding disabled button functionally using logic

savir-nosaj committed:
Update ToDoList.jsx

* destructure onCompletedTodo prop
* create new variable (filteredTodoList containing a filter, of all items with a isCompleted prop set to true
* display data depending on filter
* figured that passing key prop to ToDoListItem child is unnecessary since it’s already stored in todoList prop.
* similar with todo; there’s no need to pass specifically title. pass the entire todoList prop to child

savir-nosaj committed:
Update ToDoListItem.jsx

changes =>

* destructure onCompletedTodo prop
* li now holds a form, with checkbox, with onChange event-listener that calls onCompletedTodo, passing item key id

royemosby commented on src/App.jsx:
```diff
@@ -6,14 +6,29 @@ import { useState } from 'react';
   function handleAddTodo(title) {
     const newTodo = {
       id: Date.now(),
-      title
+      title: title,
```
Just to let you know, these lines are the equivalent.

SarahSoutoul replied:
When the key and the value are the same, you can just pass `title`, so have:

```
 const newTodo = {
      id: Date.now(),
      title: title,
      isCompleted: false
   }
```

royemosby commented on src/ToDoForm.jsx:
```diff
@@ -1,21 +1,35 @@
+    // local handleAddTodo function performs the following:
+    function handleAddTodo(title){
+        // prevent page re-render
+        title.preventDefault();
```
Using the word "title", makes it seem like `handleAddTodo` should take in text. To keep the function easy to understand, make sure that the argument name is descriptive of what the function expects. In this case, we can still use `event`.

savir-nosaj replied:
I thought it was appropriate since the input data we're sending back to `handleAddTodo` in `App`, from `ToDoForm`, is only text (todo title).
I changed it to 'title', to help me understand what information specifically, the `handleAddTodo` is expecting as the parameter.

savir-nosaj replied:
Roy, I'm mistaken. You're talking about the `handleAddTodo` function in the `ToDoForm` component.
I agree with you; the function isn't specifically taking in a string input, but rather the `event` itself.
I will fix it now!

royemosby commented on src/ToDoForm.jsx:
```diff
@@ -1,21 +1,35 @@
+                // when input is detected => run setWorkingTodo function, adding value to end of state (workingTodo) string
+                onChange={(event) => setWorkingTodoTitle(event.target.value)}
+            />
+            <button type="submit" disabled={!workingTodoTitle}>Add Todo</button>
```
👍

royemosby reviewed (changes requested):
Things look pretty good but I have a minor change request in the conversation. Also, don't forget to remove commented sections that you are no longer using.

savir-nosaj committed:
Update ToDoForm.jsx

handleAddTodo doesn't take in a string input, but rather an event.

SarahSoutoul commented on src/App.jsx:
```diff
@@ -22,6 +37,7 @@ function App() {
       />
       <ToDoList
         todoList = {todoList}
+        onCompletedTodo = {onCompleteTodo}
```
Make sure to remove spaces between your prop key and the value - it's cleaner and the best practice for props in general

SarahSoutoul commented on src/ToDoList.jsx:
```diff
@@ -1,18 +1,28 @@
+                {filteredTodoList.map((todo) => {
+                    // for each value in state arr, create instance of ToDoListItem comp., passing same props of local parent
+                    return (
+                        <ToDoListItem
```
Make sure to add a key prop to the `ToDoListItem` as  this is the element that is being returned several times as a result of the map. Like this:

`key={todo.id}`

SarahSoutoul commented on src/ToDoList.jsx:
```diff
@@ -1,18 +1,28 @@
+            <ul>
+                {filteredTodoList.map((todo) => {
+                    // for each value in state arr, create instance of ToDoListItem comp., passing same props of local parent
+                    return (
```
You wouldn't need the return here - just replace the curly brace above with parentheses and you would be fine.

SarahSoutoul commented on src/ToDoListItem.jsx:
```diff
@@ -1,6 +1,17 @@
+            <form>
+                <input
+                    type="checkbox"
+                    //checked={} - {/* add the `checked` props */} - ????????
```
Here you needed to add the following

`checked={todo.isCompleted}`

SarahSoutoul reviewed (approved):
Nice one Jason! Added a few comments but things look good.

savir-nosaj committed:
Update App.jsx

removing spaces between props key and their value

savir-nosaj committed:
Update ToDoList.jsx

- removing unnecessary return statement in map function
- replace curly bracket in .map function with parenthesis
- adding key prop to ToDoListItem

savir-nosaj committed:
Update ToDoListItem.jsx

- adding checked todo attribute, corresponding with the isCompleted prop, attached to todo
